### ISSUE TITLE
Syntax error when slack username has an apostrophe

### ISSUE THREAD

**Issue by btoconnor** [onthebeach/capistrano-slackify#38]
The curl command to slack dies when the slack username has an apostrophe in it.  I feel like it's appropriate to try and escape the name in this library.

Thoughts?

**Comment by seenmyfate** [comment]
That's an interesting one @btoconnor, I've not run into Slack usernames with an apostrophe.  It feels like a bit of an edge case - and one that could be handled by escaping the username in the configuration for those who run into this problem.  Having said that, if it is a valid username then it should really be supported without the need for any intervention, I wouldn't have any objections to using something from [`Shellwords`](http://ruby-doc.org/stdlib-2.5.0/libdoc/shellwords/rdoc/Shellwords.html#method-c-shellescape) to deal with it.

**Comment by btoconnor** [comment]
Thanks for replying.  I'm not sure I would consider the apostrophe such an edge case - they're quite common in surnames of Irish descent (and I believe other cultures as well).

Let me know if I can provide anymore information!

**Comment by seenmyfate** [comment]
Sorry, I meant more in terms of allowed characters in Slack usernames - I think I recall a restriction on Slack's side to something like letters/numbers, hyphens, full-stops and underscores, but I can't see anything obvious in the API docs to support that recollection. I wonder if it's something that changed with [the switch to display names?](https://api.slack.com/changelog/2017-09-the-one-about-usernames).  Either way, if it's causing problems then it's worth fixing - I'll gladly accept a PR or I'll aim to get to it soon.

**Comment by btoconnor** [comment]
That would make sense.  It looks like the API payload for my slack account has my "name" as "brian", but my display name as "Brian O'Connor".  The latter is what's making its way to capistrano-slackify now in our org, which is causing the failure.

**Comment by bew** [comment]
We're having the same problem: in the fields we send the commits that are being deployed, and we got commit message with a simple quote `'`, → the `curl` command fails with a syntax error..

Basically the curl command looks like:
```sh
( /usr/bin/env curl -X POST -s --data-urlencode 'payload=<the payload>' <slack hook URL> )
```
Now, if the payload contains a simple quote, the `--payload` argument will end, and the rest of the payload will be interpreted as shell arguments or commands.
I think this is a pretty bad security issue! :warning: 

Imagine a payload with something like: `dummy payload' <some url>; rm -rf ~; echo '`
This will send the payload (up to `dummy payload`) to `<some url>`, then delete the home folder of the user running the capistrano task, then print the `<slack hook URL>`... :boom:

**Comment by seenmyfate** [comment]
Given that you (as the user) control all of the inputs to the payload via configuration in your app, I don't think that scenario is likely.  Thanks for PR though @bew, I'll test that when I get the chance 👍

**Comment by bew** [comment]
> Given that you (as the user) control all of the inputs to the payload via configuration in your app, I don't think that scenario is likely.

Hmm true, then I guess we can call it is a potential risk ? or an issue waiting to happen, based on how the users will use/configure the code :smiley: 

Let me know if I can do anything, meanwhile I'll just remove that nasty quote as a workaround..

**Comment by btoconnor** [comment]
I'm not sure I agree with the assessment that this is a configuration thing that someone controls within the app.  If I were to change my name on slack to "Brian O'rm -rf /" that would be catastrophic on the server that capistrano runs on.

**Comment by cluxter** [comment]
> Given that you (as the user) control all of the inputs to the payload via configuration in your app, I don't think that scenario is likely. Thanks for PR though @bew, I'll test that when I get the chance 👍

I've got to disagree here. Nobody on Earth in 2019 should still use unescaped strings, no matter what. Decades of code injections of all sorts should have taught us that they are major attack vectors.

Escaping the string @bew is referring to should be done by _both_ the client _and_ the library.

Your library should do everything possible in its field of operation to maintain a sane environment. Same applies to the application that uses it.

You should assume input data is _always_ compromised. Never trust what goes in. A beginning of explanation here: https://en.wikipedia.org/wiki/Murphy%27s_law

And this very discussion is the best proof of what I'm saying. The author of the library thinks the client controls its input data. The developer of the client thinks the library should do (and does) the parsing job. Thus nobody does the job and it's just a matter of time before a hacker gets into a random system and starts having fun. And how did this hacker get into the said system? Thanks to these kind of flaws that nobody wanted to take care of.

Please, patch this flaw.

**Comment by seenmyfate** [comment]
@btoconnor just to confirm the username displayed in Slack by this endpoint is hardcoded in your configuration. It's the username that your bot will display (by default 'Capistrano'), nothing you do in Slack will be reflected back to this library.

@cluxter I completely agree with your view that this should be escaped, however, please remember that everything this library uses to build that payload comes from either the defaults or the configuration you add to your app.  I still don't think that the scenario @bew describes is likely, but I appreciate your concern.  This will get patched when I have the opportunity to look at #39  - feel free to take a look if you have the time.

**Comment by seenmyfate** [comment]
This now available as [v2.10.3 on Rubygems](https://rubygems.org/gems/capistrano-slackify/versions/2.10.3) - thanks for all the help 👍

**Comment by bew** [comment]
Awesome, thanks for the quick release @seenmyfate!
